README.md
# Vagrant Developer Box

| Category           | Name   |
| ------------------ | ------ |
| Operating System   | Linux Ubuntu Server 20.04 64 bit Long Term Support |
| Web Server   | Nginx 1.18.0   |
| Programming Language   | PHP 8.2.7	|
| Relational Database   | PostgreSQL 10.12   |
| Relational Database Alternative   | MySQL 8.0.33   |
| HTTP Cache        | Varnish 6.2.1 |
| In-memory Database   | Redis 6.0.6   |
| Javascript Runtime Environment   | NodeJS 18.16.1   |
| Javascript Package Manager   | NPM 9.5.1   |
| Version Control System   | Git 2.25.1   |
| PHP Dependency Manager   | Composer latest   |
| APIGen   | APIGen latest   |
| Testing Framework   | Codeception latest   |

## Installation

```bash
vagrant up // start the server
vagrant ssh // go inside the server
vagrant halt // shutdown the server
vagrant destroy // remove the server
```

## Virtual Host

For each application that you want to host, you have to create a virtual host.

You can create (copy from the example) in ``sites-available`` folder.

Then, you will have to let vagrant reload and run provisioning process again.

``vagrant provision``


# Database

You have default PostgreSQL use as follow:

* Username : devbox
* Password : secret

You have default MySQL use as follow:

* Username : root
* Password : root

You can connect for example via PGAdmin / Toad with the following information:

* Host : 192.168.33.10
* Port : 5432 (PostgreSQL)
* Port : 3306 (MySQL)

or (using SSH)

* Host : 127.0.0.1
* Port : 5433 (PostgreSQL)
* Port : 3307 (MySQL)

Please note that only in development we can connect to our virtual server from any (*) host.

## OCI8

You can also connect to Oracle database using oci8.

## Laravel

Execute this script to complete Laravel instalation. Make sure the following things:

1. PostgreSQL Database : laravel exists.
2. Redis Server service started.

```bash
composer install
composer require predis/predis
cp .env.example .env
php artisan key:generate
nano .env
	DB_CONNECTION=pgsql
	DB_PORT=5432
	DB_DATABASE=laravel
	DB_USERNAME=devbox
	DB_PASSWORD=secret

	CACHE_DRIVER=redis
	SESSION_DRIVER=redis
php artisan migrate
php artisan db:seed

# Below are optional

# Unix
npm install
# Windows
npm install --no-bin-links
npm run dev
```

## Login

Username (email) : check laravel database on table users.
Password : password

## Theme

SBAdmin 2

Check https://github.com/davidpramana/laravel8-sbadmin2

## Provision.sh

This is the default ``provision.sh``

```bash
main() {
	update_go
	locale_go
	common_go
	npm_go
	php_go
	redis_go
	# codeception_go
	composer_go
	# apigen_go
	nginx_go
	postgresql_go
	# mysql_go
	# oci8_go
	varnish_go
	autoremove_go
	complete_go
}
```

Please change the ``main()`` function to the following after provisioning for the first time:


```bash
main() {
	# update_go
	# locale_go
	# common_go
	# npm_go
	# php_go
	# redis_go
	# codeception_go
	# composer_go
	# apigen_go
	nginx_go
	# postgresql_go
	# mysql_go
	# oci8_go
	# varnish_go
	autoremove_go
	complete_go
}
```

Basically only set nginx for daily development.

## Vagrant VirtualBox Guess Addition

Jika menemukan warning seperti ini:

```
The guest additions on this VM do not match the installed version of
VirtualBox! In most cases this is fine, but in rare cases it can
shared folder errors, please make sure the guest additions within the
virtual machine match the version of VirtualBox you have installed on
your host and reload your VM.

Guest Additions Version: 6.1.38
VirtualBox Version: 7.0
```

Maka eksekusi command berikut:

```
vagrant vbguest --auto-reboot --no-provision
```

CHANGELOG.md
# CHANGELOG

## 0.6.2

- MODIFIED; OS Ubuntu 20.04 to 22.04 LTS (Jammy)
- ADDED; Custom PHP-FPM Config
- ADDED; MySQL Tuner Perl Script
- MODIFIED; Redis PPA to Official RedisLab
- MODIFIED; PostgreSQL 10 to 15

## 0.6.1

- MODIFIED; OS Ubuntu 22.04 to 20.04 LTS (Focal)

## 0.6.0

- MODIFIED; OS to Ubuntu 22.04 LTS (Jammy)
- MODIFIED; PHP 8.0 to 8.2
- MODIFIED; NodeJS 14 to 18
- MODIFIED; README.md for current technology stack & version
- ADDED; Support for VirtualBox Guest Additions plugin disable auto update
- ADDED; Connect using cable for network

## 0.5.0

- MODIFIED; OS to Ubuntu 20.04 LTS (Focal)
- MODIFIED; OCI8 PHP 7.3 to 8.0
- ADDED; Support for Windows and Unix VM synced folder
- MODIFIED; README.md for current technology stack & version

## 0.4.1

- MODIFIED; Redis Server installed as service

## 0.4.0

- MODIFIED; default example-web.conf nginx
- ADDED; Timezone to Asia/Jakarta
- ADDED; complete_go function to upgrade and make sure varnish & nginx works
- MODIFIED; nginx_go to remove default site
- ADDED; Remove varnish template files on complete_go
- MODIFIED; Change NodeJS version from 10 to 14 LTS
- ADDED; php-redis for PHP
- MODIFIED; Shared Folder type to NFS with options (nolock)
- ADDED; Redis Server installed as service
- MODIFIED; Change PHP 7.3 to 8.0

## 0.3.0

- ADDED; php7.3-redis for Laravel >= 6
- ADDED; unzip for common packages
- ADDED; redis-server start on startup
- ADDED; webpack-cli and yarn for npm global packages
- MODIFIED; NPM 10 to 14 LTS
- ADDED; Set timezone to Asia/Jakarta

## 0.2.1

- MODIFIED; Steps to install PosgreSQL
- REMOVED; Packages python-psycopg2 from PostgreSQL

## 0.2.0

- ADDED; Locale english base
- ADDED; zip and software-properties-common for common module
- REMOVED; NPM Strongloop, Lodash, Graceful FS and Sails as default node modules
- MODIFIED; module call priorities
- MODIFIED; Seperate composer, codeception, and apigen module
- MODIFIED; Remove existing nginx site available in guest before mimicing host sites available
- ADDED; OCI8 for PHP and Oracle Client 12.1 (basic and development)
- MODIFIED; Nginx server block from using php5-fpm to php7.0-fpm
- MODIFIED; Nginx server port from 80 to 8080 because of Varnish
- ADDED; Varnish as HTTP Cache
- MODIFIED; Vagrant file ports
- MODIFIED; README.md
- ADDED; CHANGELOG.md and CONTRIBUTORS.md

## 0.1.0

- Initial release

example.provision.sh
#!/bin/bash

# This function is called at the very bottom of the file
main() {
	update_go
	locale_go
	common_go
	npm_go
	php_go
	redis_go
	# codeception_go
	composer_go
	# apigen_go
	nginx_go
	postgresql_go
	# mysql_go
	# oci8_go
	varnish_go
	autoremove_go
	complete_go
}

# GO MODULES

update_go() {
	# Update the server
	apt-get update
}

locale_go() {
	# Set TimeZone
	timedatectl set-timezone Asia/Jakarta

	# install language pack english base
	apt-get -y install language-pack-en-base
	# set locale
	if grep -Fxq "LC_ALL=en_US.UTF-8" /etc/default/locale; then
		# code if found : ignore
		echo 'ignore'
	else
		# code if not found : append file
		sed -i -e "s/LC_ALL=/LC_ALL=en_US.UTF-8/g" /etc/default/locale
	fi

	# set language
	if grep -Fxq "LANG=en_US.UTF-8" /etc/default/locale; then
		# code if found : ignore
		echo 'ignore'
	else
		# code if not found : append file
		sed -i -e "s/LANG=/LANG=en_US.UTF-8/g" /etc/default/locale
	fi

	# reconfigure locales
	export LC_ALL=en_US.UTF-8
	export LANG=en_US.UTF-8
	dpkg-reconfigure locales
}

common_go() {
	# Install common softwares
	apt-get -y install mcrypt git curl build-essential zip software-properties-common gcc g++ make libpng-dev unzip
}

npm_go() {
	curl -sL https://deb.nodesource.com/setup_18.x | sudo -E bash -
	apt-get -y install nodejs
	npm install -g gulp-cli gulp pm2 cross-env
}

autoremove_go() {
	apt-get -y autoremove
}

nginx_go() {
	# Install Nginx
	apt-get -y install nginx

	# Remove default site
	rm -f /etc/nginx/sites-available/default
	rm -f /etc/nginx/sites-enabled/default

	# Remove existing virtual hosts
	rm -f /etc/nginx/sites-enabled/*.conf

	# Server Block / Virtual Host
	for site in ./sites-available/*
	do
		echo 'Processing ' $(basename $site) ' file...'
		cp /vagrant/sites-available/$(basename $site) /etc/nginx/sites-available
		rm -f /etc/nginx/sites-enabled/$(basename $site)
		ln -s /etc/nginx/sites-available/$(basename $site) /etc/nginx/sites-enabled/
	done

	nginx_handler
}

php_go() {
	# Add PHP Repository
	add-apt-repository ppa:ondrej/php

	# Update Repository
	apt-get update

	# Install PHP 8
	apt-get install -y php8.2 php8.2-bcmath php8.2-bz2 php8.2-cgi php8.2-cli php8.2-common php8.2-curl php8.2-dba php8.2-enchant php8.2-fpm php8.2-gd php8.2-gmp php8.2-imap php8.2-interbase php8.2-intl php8.2-ldap php8.2-mbstring php8.2-mysql php8.2-odbc php8.2-opcache php8.2-pgsql php8.2-phpdbg php8.2-pspell php8.2-readline php8.2-snmp php8.2-soap php8.2-sqlite3 php8.2-sybase php8.2-tidy php8.2-xml php8.2-xmlrpc php8.2-xsl php8.2-zip php8.2-redis

	# Modify PHP FPM Config
	sed -i -e "s/pm = dynamic/pm = static/g" /etc/php/8.2/fpm/pool.d/www.conf
	sed -i -e "s/pm.max_children = 5/pm.max_children = 44/g" /etc/php/8.2/fpm/pool.d/www.conf
	sed -i -e "s/pm.start_servers = 2/pm.start_servers = 11/g" /etc/php/8.2/fpm/pool.d/www.conf
	sed -i -e "s/pm.min_spare_servers = 1/pm.min_spare_servers = 11/g" /etc/php/8.2/fpm/pool.d/www.conf
	sed -i -e "s/pm.max_spare_servers = 3/pm.max_spare_servers = 33/g" /etc/php/8.2/fpm/pool.d/www.conf

	nginx_handler
}

redis_go() {
	# Add Redis PPA
	add-apt-repository ppa:redislabs/redis

	# Update Repository
	apt-get update

	# Install Redis Server
	apt-get -y install redis-server

	# Install as Service
	sed -i "s/supervised no/supervised systemd/g" /etc/redis/redis.conf
	service redis-server restart
}

codeception_go() {
	# Install Codeception globally
	if [ ! -f "/usr/local/bin/phpunit" ]; then
		curl -O -L http://codeception.com/codecept.phar
		chmod +x codecept.phar
		mv codecept.phar /usr/local/bin/codecept
	fi
}

composer_go() {
	# Install latest version of Composer globally
	if [ ! -f "/usr/local/bin/composer" ]; then
		curl -O -L https://getcomposer.org/composer.phar
		chmod +x composer.phar
		mv composer.phar /usr/local/bin/composer
	fi
}

apigen_go() {
	# Install APIGen
	if [ ! -f "/usr/local/bin/apigen" ]; then
		curl -O -L http://apigen.org/apigen.phar
		chmod +x apigen.phar
		mv apigen.phar /usr/local/bin/apigen
	fi
}

postgresql_go() {

	# Get PostgreSQL key
	wget -q https://www.postgresql.org/media/keys/ACCC4CF8.asc -O- | sudo apt-key add -

	# Add the repository to the apt sources
	echo "deb http://apt.postgresql.org/pub/repos/apt/ jammy-pgdg main" > /etc/apt/sources.list.d/PostgreSQL.list

	# Update apt
	apt-get update

	# Install PostgreSQL
	apt-get -y install postgresql-15 #python-psycopg2

	# create pg cluster
	pg_createcluster 15 main --start

	# Allow access from Host
	echo "host all all all password" >> /etc/postgresql/15/main/pg_hba.conf
	sed -i -e "s/#listen_addresses = 'localhost'/listen_addresses = '*'/g" /etc/postgresql/15/main/postgresql.conf

	# run
	service postgresql start

	# create default user
	sudo -u postgres bash -c "psql -c \"create user devbox with password 'secret' CREATEDB;\""

	# Add Custom PostgreSQL Config
	cp -f ./postgresql-custom.conf /etc/postgresql/15/main/conf.d/custom.conf

	# Remove PostgreSQL Custom Config Templates
	rm /home/vagrant/postgresql-custom.conf

	# PostgreSQL Handler
	postgresql_handler
}

mysql_go() {
	# Set Default root Password
	echo "mysql-server mysql-server/root_password password root" | sudo debconf-set-selections
	echo "mysql-server mysql-server/root_password_again password root" | sudo debconf-set-selections

	# Install MySQL
	apt-get -y install mysql-server

	# Download MySQL Tuner
	wget http://mysqltuner.pl/ -O /home/vagrant/mysqltuner.pl
	wget https://raw.githubusercontent.com/major/MySQLTuner-perl/master/basic_passwords.txt -O basic_passwords.txt
	wget https://raw.githubusercontent.com/major/MySQLTuner-perl/master/vulnerabilities.csv -O vulnerabilities.csv
	chmod +x /home/vagrant/mysqltuner.pl

	# run
	service mysql start
}

oci8_go() {
	# install OCI8
	apt-get -y install libaio1 build-essential php-pear php8.2-dev

	# Set executable
	chmod +x *.deb

	# install oracle instant client12.1 basic
	dpkg -i ./oracle-instantclient12.1-basic_12.1.0.2.0-2_amd64.deb

	# install oracle instant client12.1 dev
	dpkg -i ./oracle-instantclient12.1-devel_12.1.0.2.0-2_amd64.deb

	# install pecl oci8 (and ignore error when installing twice)
	yes 'instantclient,/usr/lib/oracle/12.1/client64/lib' | pecl install oci8 || true

	# add oci8 extension to php.ini (CLI)
	if grep -Fxq "extension=oci8.so" /etc/php/8.2/cli/php.ini; then
		# code if found : ignore
		echo 'extension oci8 already exists in cli'
	else
		# code if not found : append file
		echo 'extension=oci8.so' >> /etc/php/8.2/cli/php.ini
	fi

	# add oci8 extension to php.ini (FPM)
	if grep -Fxq "extension=oci8.so" /etc/php/8.2/fpm/php.ini; then
		# code if found : ignore
		echo 'extension oci8 already exists in fpm'
	else
		# code if not found : append file
		echo 'extension=oci8.so' >> /etc/php/8.2/fpm/php.ini
	fi
}

varnish_go() {
	# install apt https
	apt-get -y install curl gnupg apt-transport-https

	# Get Varnish key & repository source
	curl -s https://packagecloud.io/install/repositories/varnishcache/varnish60lts/script.deb.sh | sudo bash

	# Update apt
	apt-get update

	# Install Varnish
	apt-get -y install varnish

	# Modify Varnish port
	cp -f ./varnish /etc/default/varnish

	# Modify Varnish VCL
	cp -f ./varnish.vcl /etc/varnish/default.vcl

	# Modify Varnish service
	cp -f ./varnish.service /lib/systemd/system/varnish.service
	systemctl daemon-reload

	# Remove Varnish File Templates
	rm /home/vagrant/varnish
	rm /home/vagrant/varnish.service
	rm /home/vagrant/varnish.vcl

	# Ensure Varnish is started
	varnish_handler

	# restart nginx
	nginx_handler
}

complete_go() {
	# upgrade packages
	apt-get -y upgrade

	# Set TimeZone
	timedatectl set-timezone Asia/Jakarta

	# restart nginx
	nginx_handler

	# Ensure Varnish is started
	varnish_handler

	# Change .config folder ownership
	chown -R vagrant:vagrant /home/vagrant/.config

	# Add Vagrant to www-data Nginx
	usermod -a -G www-data vagrant

	# Remove Varnish File Templates
	rm /home/vagrant/varnish
	rm /home/vagrant/varnish.service
	rm /home/vagrant/varnish.vcl

	# Remove PostgreSQL Custom Config Templates
	rm /home/vagrant/postgresql-custom.conf
}

# HANDLER

nginx_handler() {
	service nginx restart
	service php8.2-fpm restart
}

varnish_handler() {
	service varnish restart
}

postgresql_handler() {
	service postgresql restart
}

main
exit 0

provision.sh
#!/bin/bash

# This function is called at the very bottom of the file
main() {
	# update_go
	# locale_go
	# common_go
	# npm_go
	# php_go
	# redis_go
	# codeception_go
	# composer_go
	# apigen_go
	nginx_go
	# postgresql_go
	# mysql_go
	# oci8_go
	# varnish_go
	autoremove_go
	complete_go
}

# GO MODULES

update_go() {
	# Update the server
	apt-get update
}

locale_go() {
	# Set TimeZone
	timedatectl set-timezone Asia/Jakarta

	# install language pack english base
	apt-get -y install language-pack-en-base
	# set locale
	if grep -Fxq "LC_ALL=en_US.UTF-8" /etc/default/locale; then
		# code if found : ignore
		echo 'ignore'
	else
		# code if not found : append file
		sed -i -e "s/LC_ALL=/LC_ALL=en_US.UTF-8/g" /etc/default/locale
	fi

	# set language
	if grep -Fxq "LANG=en_US.UTF-8" /etc/default/locale; then
		# code if found : ignore
		echo 'ignore'
	else
		# code if not found : append file
		sed -i -e "s/LANG=/LANG=en_US.UTF-8/g" /etc/default/locale
	fi

	# reconfigure locales
	export LC_ALL=en_US.UTF-8
	export LANG=en_US.UTF-8
	dpkg-reconfigure locales
}

common_go() {
	# Install common softwares
	apt-get -y install mcrypt git curl build-essential zip software-properties-common gcc g++ make libpng-dev unzip
}

npm_go() {
	curl -sL https://deb.nodesource.com/setup_18.x | sudo -E bash -
	apt-get -y install nodejs
	npm install -g gulp-cli gulp pm2 cross-env
}

autoremove_go() {
	apt-get -y autoremove
}

nginx_go() {
	# Install Nginx
	apt-get -y install nginx

	# Remove default site
	rm -f /etc/nginx/sites-available/default
	rm -f /etc/nginx/sites-enabled/default

	# Remove existing virtual hosts
	rm -f /etc/nginx/sites-enabled/*.conf

	# Server Block / Virtual Host
	for site in ./sites-available/*
	do
		echo 'Processing ' $(basename $site) ' file...'
		cp /vagrant/sites-available/$(basename $site) /etc/nginx/sites-available
		rm -f /etc/nginx/sites-enabled/$(basename $site)
		ln -s /etc/nginx/sites-available/$(basename $site) /etc/nginx/sites-enabled/
	done

	nginx_handler
}

php_go() {
	# Add PHP Repository
	add-apt-repository ppa:ondrej/php

	# Update Repository
	apt-get update

	# Install PHP 8
	apt-get install -y php8.2 php8.2-bcmath php8.2-bz2 php8.2-cgi php8.2-cli php8.2-common php8.2-curl php8.2-dba php8.2-enchant php8.2-fpm php8.2-gd php8.2-gmp php8.2-imap php8.2-interbase php8.2-intl php8.2-ldap php8.2-mbstring php8.2-mysql php8.2-odbc php8.2-opcache php8.2-pgsql php8.2-phpdbg php8.2-pspell php8.2-readline php8.2-snmp php8.2-soap php8.2-sqlite3 php8.2-sybase php8.2-tidy php8.2-xml php8.2-xmlrpc php8.2-xsl php8.2-zip php8.2-redis

	# Modify PHP FPM Config
	sed -i -e "s/pm = dynamic/pm = static/g" /etc/php/8.2/fpm/pool.d/www.conf
	sed -i -e "s/pm.max_children = 5/pm.max_children = 44/g" /etc/php/8.2/fpm/pool.d/www.conf
	sed -i -e "s/pm.start_servers = 2/pm.start_servers = 11/g" /etc/php/8.2/fpm/pool.d/www.conf
	sed -i -e "s/pm.min_spare_servers = 1/pm.min_spare_servers = 11/g" /etc/php/8.2/fpm/pool.d/www.conf
	sed -i -e "s/pm.max_spare_servers = 3/pm.max_spare_servers = 33/g" /etc/php/8.2/fpm/pool.d/www.conf

	nginx_handler
}

redis_go() {
	# Add Redis PPA
	add-apt-repository ppa:redislabs/redis

	# Update Repository
	apt-get update

	# Install Redis Server
	apt-get -y install redis-server

	# Install as Service
	sed -i "s/supervised no/supervised systemd/g" /etc/redis/redis.conf
	service redis-server restart
}

codeception_go() {
	# Install Codeception globally
	if [ ! -f "/usr/local/bin/phpunit" ]; then
		curl -O -L http://codeception.com/codecept.phar
		chmod +x codecept.phar
		mv codecept.phar /usr/local/bin/codecept
	fi
}

composer_go() {
	# Install latest version of Composer globally
	if [ ! -f "/usr/local/bin/composer" ]; then
		curl -O -L https://getcomposer.org/composer.phar
		chmod +x composer.phar
		mv composer.phar /usr/local/bin/composer
	fi
}

apigen_go() {
	# Install APIGen
	if [ ! -f "/usr/local/bin/apigen" ]; then
		curl -O -L http://apigen.org/apigen.phar
		chmod +x apigen.phar
		mv apigen.phar /usr/local/bin/apigen
	fi
}

postgresql_go() {

	# Get PostgreSQL key
	wget -q https://www.postgresql.org/media/keys/ACCC4CF8.asc -O- | sudo apt-key add -

	# Add the repository to the apt sources
	echo "deb http://apt.postgresql.org/pub/repos/apt/ jammy-pgdg main" > /etc/apt/sources.list.d/PostgreSQL.list

	# Update apt
	apt-get update

	# Install PostgreSQL
	apt-get -y install postgresql-15 #python-psycopg2

	# create pg cluster
	pg_createcluster 15 main --start

	# Allow access from Host
	echo "host all all all password" >> /etc/postgresql/15/main/pg_hba.conf
	sed -i -e "s/#listen_addresses = 'localhost'/listen_addresses = '*'/g" /etc/postgresql/15/main/postgresql.conf

	# run
	service postgresql start

	# create default user
	sudo -u postgres bash -c "psql -c \"create user devbox with password 'secret' CREATEDB;\""

	# Add Custom PostgreSQL Config
	cp -f ./postgresql-custom.conf /etc/postgresql/15/main/conf.d/custom.conf

	# Remove PostgreSQL Custom Config Templates
	rm /home/vagrant/postgresql-custom.conf

	# PostgreSQL Handler
	postgresql_handler
}

mysql_go() {
	# Set Default root Password
	echo "mysql-server mysql-server/root_password password root" | sudo debconf-set-selections
	echo "mysql-server mysql-server/root_password_again password root" | sudo debconf-set-selections

	# Install MySQL
	apt-get -y install mysql-server

	# Download MySQL Tuner
	wget http://mysqltuner.pl/ -O /home/vagrant/mysqltuner.pl
	wget https://raw.githubusercontent.com/major/MySQLTuner-perl/master/basic_passwords.txt -O basic_passwords.txt
	wget https://raw.githubusercontent.com/major/MySQLTuner-perl/master/vulnerabilities.csv -O vulnerabilities.csv
	chmod +x /home/vagrant/mysqltuner.pl

	# run
	service mysql start
}

oci8_go() {
	# install OCI8
	apt-get -y install libaio1 build-essential php-pear php8.2-dev

	# Set executable
	chmod +x *.deb

	# install oracle instant client12.1 basic
	dpkg -i ./oracle-instantclient12.1-basic_12.1.0.2.0-2_amd64.deb

	# install oracle instant client12.1 dev
	dpkg -i ./oracle-instantclient12.1-devel_12.1.0.2.0-2_amd64.deb

	# install pecl oci8 (and ignore error when installing twice)
	yes 'instantclient,/usr/lib/oracle/12.1/client64/lib' | pecl install oci8 || true

	# add oci8 extension to php.ini (CLI)
	if grep -Fxq "extension=oci8.so" /etc/php/8.2/cli/php.ini; then
		# code if found : ignore
		echo 'extension oci8 already exists in cli'
	else
		# code if not found : append file
		echo 'extension=oci8.so' >> /etc/php/8.2/cli/php.ini
	fi

	# add oci8 extension to php.ini (FPM)
	if grep -Fxq "extension=oci8.so" /etc/php/8.2/fpm/php.ini; then
		# code if found : ignore
		echo 'extension oci8 already exists in fpm'
	else
		# code if not found : append file
		echo 'extension=oci8.so' >> /etc/php/8.2/fpm/php.ini
	fi
}

varnish_go() {
	# install apt https
	apt-get -y install curl gnupg apt-transport-https

	# Get Varnish key & repository source
	curl -s https://packagecloud.io/install/repositories/varnishcache/varnish60lts/script.deb.sh | sudo bash

	# Update apt
	apt-get update

	# Install Varnish
	apt-get -y install varnish

	# Modify Varnish port
	cp -f ./varnish /etc/default/varnish

	# Modify Varnish VCL
	cp -f ./varnish.vcl /etc/varnish/default.vcl

	# Modify Varnish service
	cp -f ./varnish.service /lib/systemd/system/varnish.service
	systemctl daemon-reload

	# Remove Varnish File Templates
	rm /home/vagrant/varnish
	rm /home/vagrant/varnish.service
	rm /home/vagrant/varnish.vcl

	# Ensure Varnish is started
	varnish_handler

	# restart nginx
	nginx_handler
}

complete_go() {
	# upgrade packages
	apt-get -y upgrade

	# Set TimeZone
	timedatectl set-timezone Asia/Jakarta

	# restart nginx
	nginx_handler

	# Ensure Varnish is started
	varnish_handler

	# Change .config folder ownership
	chown -R vagrant:vagrant /home/vagrant/.config

	# Add Vagrant to www-data Nginx
	usermod -a -G www-data vagrant

	# Remove Varnish File Templates
	rm /home/vagrant/varnish
	rm /home/vagrant/varnish.service
	rm /home/vagrant/varnish.vcl

	# Remove PostgreSQL Custom Config Templates
	rm /home/vagrant/postgresql-custom.conf
}

# HANDLER

nginx_handler() {
	service nginx restart
	service php8.2-fpm restart
}

varnish_handler() {
	service varnish restart
}

postgresql_handler() {
	service postgresql restart
}

main
exit 0